cumulus/client/consensus/aura/src/collators/slot_based/slot_timer.rs

```diff
@@ -173,6 +191,115 @@ where
 		))
 	}
 
+	/// Compute the time until the next slot changes.
+	fn compute_time_until_next_slot_change(&mut self) -> Result<(Duration, Slot), ()> {
+		let Ok(slot_duration) = crate::slot_duration(&*self.client) else {
+			tracing::error!(target: LOG_TARGET, "Failed to fetch slot duration from runtime.");
+			return Err(())
+		};
+
+		let now = duration_now();
+		let now = now.as_millis().saturating_sub(self.time_offset.as_millis());
+
+		let last_reported_slot = self.last_reported_slot.unwrap_or_default();
+		let Some(last_slot_timestamp) = last_reported_slot.timestamp(slot_duration) else {
+			tracing::error!(target: LOG_TARGET, "Failed to obtain the last slot timestamp");
+			return Err(())
+		};
+
+		// Compute when the next different slot starts.
+		let next_different_slot_time =
+			last_slot_timestamp.as_millis() + slot_duration.as_millis() as u64;
+		let remaining_millis = next_different_slot_time.saturating_sub(now as u64);
+		let next_aura_slot =
+			Slot::from_timestamp(Timestamp::from(next_different_slot_time as u64), slot_duration);
+
+		Ok((Duration::from_millis(remaining_millis as u64), next_aura_slot))
+	}
+
+	/// Adjust the authoring duration to fit into the slot timing.
+	///
+	/// Returns the adjusted authoring duration and the slot that it corresponds to.
+	pub fn adjust_authoring_duration(
+		&mut self,
+		mut authoring_duration: Duration,
+	) -> Option<Duration> {
+		let Ok((duration, next_slot)) = self.time_until_next_slot() else {
```

Comment:
We should rename `time_until_next_slot`, because its not really reflecting the next slot, but more the next block timepoint.